Source/WebCore/Modules/fetch/FetchHeaders.cpp

```diff
@@ -225,19 +277,62 @@ void FetchHeaders::filterAndFill(const HTTPHeaderMap& headers, Guard guard)
     }
 }
 
+inline bool compareIteratorKeys(const String& a, const String& b)
+{
+    // null in the iterator's m_keys represents Set-Cookie.
+    if (a.isNull() && b.isNull())
+        return false;
+    return WTF::codePointCompareLessThan(
+        a.isNull() ? "set-cookie"_s : a,
+        b.isNull() ? "set-cookie"_s : b);
+}
+
 std::optional<KeyValuePair<String, String>> FetchHeaders::Iterator::next()
 {
     if (m_keys.isEmpty() || m_updateCounter != m_headers->m_updateCounter) {
+        bool hasSetCookie = !m_headers->m_setCookieValues.isEmpty();
         m_keys.resize(0);
-        m_keys.reserveCapacity(m_headers->m_headers.size());
-        for (auto& header : m_headers->m_headers)
+        m_keys.reserveCapacity(m_headers->m_headers.size() + (hasSetCookie ? 1 : 0));
+        for (auto& header : m_headers->m_headers) {
+            ASSERT(!header.key.isNull());
             m_keys.uncheckedAppend(header.key.convertToASCIILowercase());
-        std::sort(m_keys.begin(), m_keys.end(), WTF::codePointCompareLessThan);
+        }
+        if (hasSetCookie)
+            m_keys.uncheckedAppend(String());
+        std::sort(m_keys.begin(), m_keys.end(), compareIteratorKeys);
+
+        // We adjust the current index to work with Set-Cookie headers.
+        // This relies on the fact that `m_currentIndex + m_setCookieIndex`
+        // gives you the current total index into the iteration.
+        m_currentIndex += m_setCookieIndex;
+        if (hasSetCookie) {
+            size_t setCookieKeyIndex = std::lower_bound(m_keys.begin(), m_keys.end(), String(), compareIteratorKeys) - m_keys.begin();
+            if (m_currentIndex < setCookieKeyIndex)
+                m_setCookieIndex = 0;
+            else {
+                m_setCookieIndex = std::min(m_currentIndex - setCookieKeyIndex, m_headers->m_setCookieValues.size());
+                m_currentIndex -= m_setCookieIndex;
+            }
+        } else
+            m_setCookieIndex = 0;
+
         m_updateCounter = m_headers->m_updateCounter;
     }
     while (m_currentIndex < m_keys.size()) {
-        auto key = m_keys[m_currentIndex++];
-        auto value = m_headers->m_headers.get(key);
+        auto key = m_keys[m_currentIndex];
+        String value;
+        if (key.isNull()) {
```

Comment:
We usually tend to use return  and  so on to reduce if / else.
That would mean something like:
```
if (key.isNull()) {
    if (m_setCookieIndex < m_headers->m_setCookieValues.size())
        return KeyValuePair<String, String> { ..., ...};
    m_currentIndex++;
    continue;
}
m_currentIndex++;
...

```